### PR TITLE
feat(codebuild): add filter groups feature for GitHub, GitHubEnterpri…

### LINKED ISSUE #1842: Allow to set FilterGroups for Build Triggers in CodeBuild Project
Looks like there's the `Webhook` property defined in the Layer 1 resource. According to the documentation, it should be possible to set `FilterGroups` as well - https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codebuild-project-projecttriggers.html#cfn-codebuild-project-projecttriggers-filtergroups

I was trying to fall back to L1 but it's missing there.

```typescript
const projectResource = project.node.findChild(
      "Resource"
    ) as codebuild.CfnProject;

 projectResource.propertyOverrides.triggers = { // doesn't take `FilterGroups'
    webhook: true
}
```

Would appreciate hints on how to make it work.

### PR DESCRIPTION
**Pull request by Kaixiang-AWS**
…se and BitBucket

Fixes #1842

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-codebuild/lib/source.ts [hunk 1]
```diff
@@ -106,6 +105,67 @@ export abstract class GitBuildSource extends BuildSource {
...
+   * Whether to send notifications on your build's start and end.
+   *
+   * @default true
+   */
+  readonly reportBuildStatus?: boolean;
+
+  /**
+   * A list of lists of WebhookFilter objects used to determine which webhook events are triggered.
+   */
+  readonly filterGroups?: WebhookFilter[][];
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/source.ts]: Why a list of lists? What are the semantics here?

File: packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts [hunk 2]
```diff
@@ -1153,6 +1232,28 @@ export = {
...
+          repo: 'testrepo',
+          cloneDepth: 3,
+          filterGroups: [
+            [
+              { type: WebhookFilterType.Event, pattern: 'PUSH' }
+            ]
+          ],
+        })
+      });
+    }, Error, "filterGroups property could only be set when webhook property is true");
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts]: I would actually change this behavior. I think if `filterGroups` was specified, and `webhook` was not, I think `webhook` should default to `true`. Only specifying `filterGroups` **and** `webhook: false` should result in an exception.

What do you think?

File: packages/@aws-cdk/aws-codebuild/lib/source.ts [hunk 3]
```diff
@@ -334,6 +366,41 @@ export class BitBucketSource extends GitBuildSource {
...
+   * @default false
+   */
+  readonly excludeMatchedPattern?: boolean,
+}
+
+/**
+ * Filter types for webhook filters
+ */
+export enum WebhookFilterType {
+  Event = 'EVENT',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/source.ts]: What does `Event` exactly mean? Are there some limits to what can `pattern` be when `type` is `Event`? If so, I think we might be missing a modeling opportunity here...

File: packages/@aws-cdk/aws-codebuild/lib/source.ts (hunk 1 above)
Comment (Kaixiang-AWS) [reply packages/@aws-cdk/aws-codebuild/lib/source.ts]: It's an OR operation between different lists and an AND operation inside the list.
For example, if I have filter groups `[[A,B],[C,D]]`, build will only be triggered if `(A&B)|(C&D)`

File: packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts (hunk 2 above)
Comment (Kaixiang-AWS) [reply packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts]: I'm thinking it in a different way. If customers want to remove filter groups but keep webhook enabled, they don't need to add `webhook: true` which is a cognitive load for customers.

File: packages/@aws-cdk/aws-codebuild/lib/source.ts (hunk 3 above)
Comment (Kaixiang-AWS) [reply packages/@aws-cdk/aws-codebuild/lib/source.ts]: Different from other types, pattern for `Event` could only be one or more actions concatenated with `,`. I couldn't think of a good model to model this behavior. Please let me know if you have some idea in mind.

Refs:
[filterGroups API ref](https://docs.aws.amazon.com/codebuild/latest/APIReference/API_CreateWebhook.html#CodeBuild-CreateWebhook-request-filterGroups)
[filterGroups user guide](https://docs.aws.amazon.com/codebuild/latest/userguide/sample-github-pull-request.html)

File: packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts]: Right. So when a dilemma like this happens, I like to compare the number of customers in each case. Here, you have every single person who wants to use filter groups, vs. a group that first uses filter groups, and then stops using them. It's pretty clear that the first group is larger (as the second group is a subset of the first one), so I think it makes more sense to optimize the experience for the larger group.

Comment (Kaixiang-AWS) [reply packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts]: I think we should follow the principle of least astonishment here. Customers won't be surprised to see an error if they specify filterGroups but don't specify webhook. However, they will feel surprised and confused when removing filterGroups causes their webhook being deleted. Could you please give me an example of somewhere else doing this kind of override behavior to help me better understand this case? I couldn't think of any at this moment.

Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts]: > Could you please give me an example of somewhere else doing this kind of override behavior to help me better understand this case? I couldn't think of any at this moment.

Sure. As an example, in CodeDeploy's `ServerDeploymentGroup`, we change the value of the rollback alarm configuration depending on whether the customer has provided any alarms (see [here](https://github.com/awslabs/aws-cdk/blob/ed39a8c4c00d5261bf6deb17134ed3f81d75cf63/packages/%40aws-cdk/aws-codedeploy/lib/rollback-config.ts#L23) and [here](https://github.com/awslabs/aws-cdk/blob/ed39a8c4c00d5261bf6deb17134ed3f81d75cf63/packages/%40aws-cdk/aws-codedeploy/lib/server/deployment-group.ts#L214)).

**Comment by skinny85** [comment]
Hey @Kaixiang-AWS ,

I've uploaded my proposal on how I see this API. The usage looks like as follows:

```ts
new codebuild.Project(stack, 'Project', {
  source: new codebuild.BitBucketSource({
    // ...
    filterGroups: [
      codebuild.FilterGroup.inEventOf(codebuild.EventAction.PUSH).andBranchIs('master'),
      codebuild.FilterGroup.inEventOf(codebuild.EventAction.PULL_REQUEST_CREATED).andBaseBranchIs('master'),
    ],
  })
});
```

I'm curious what you think of doing it this way. Let me know!

Thanks,
Adam

**Comment by skinny85** [comment]
Updated with the proposed final version, let me know what you think @Kaixiang-AWS !

**Comment by skinny85** [comment]
Rebased on top of `0.30.0`.

**Comment by Kaixiang-AWS** [comment]
LGTM.

**Pull request merged**